### PR TITLE
Improved relationship serializer options

### LINKED ISSUE #14: has_many serializer inference
Serializer inference on relationships can be improved.
For instance `belongs_to :owner, record_type: :user` uses `OwnerSerializer`.
In my opinion, it would make more sens to use `UserSerializer`.

I have a patch here: https://github.com/ThomasSevestre/fast_jsonapi/commit/46b5fdefb675159d89ec01b66cb45dea1590cbdf

It is a breaking change, I'll make a PR if you are interested.

### LINKED ISSUE #29: Allow a relationship's serializer to be computed based on the object
Thanks so much for picking up the ball with this project!  I was an early contributor.

When a relationship is defined, the serializer needs to determine which serializer to use.  It uses both the optional `serializer` and `polymorphic` options to help guide it.  It gets the job done in most cases, but there are definitely some hard rules that don't necessarily apply in all cases.

Consider the case of Single Table Inheritance.  Say you have an `Animal` base class and 20 subclasses, `Dog`, `Pig`, `Giraffe`, etc.  You may have different serialization rules for each, so they should each have their own serializer.  Or maybe they all have the same serializer.  Or maybe they have the same base class serializer, and only the ones that have something custom would need to define their own serializer.  Maybe this can be done today using a `polymorphic` hash, but (a) you would have to list every possible class in the hash, and (b) it looks like there are some similar rules and limitations with how `record_type` is defined with polymorphic hashes.

I think an elegant solution would be to simply allow the `serializer` option to be specified as a `Proc`, like we do with attributes, conditionals, etc.  It can include the same `object` and `params` arguments.  This way, the author has total control over selecting the serializer based on the actual object that it is attempting to serialize.

What do you think of this approach?  If it looks good, I will implement and create a PR.

Thanks again!

### LINKED ISSUE #30: Polymorphic hash does not select the proper serializer for included records
While reviewing the code and implementing #29, I noticed that there are no tests for specifying `polymorphic` as a hash, and I can't seem to find any code to support the implementation.

This is the code for determining the serializer for a specific record within a relationship (from [here](https://github.com/fast-jsonapi/fast_jsonapi/blob/1d7c18f5dac0177a0a4d8728d84b4fd63b2d91df/lib/fast_jsonapi/serialization_core.rb#L139)):

```ruby
if relationship_item.polymorphic.is_a?(Hash)
  record_type = inc_obj.class.name.demodulize.underscore
  serializer = self.compute_serializer_name(inc_obj.class.name.demodulize.to_sym).to_s.constantize
end
```

And [`compute_serializer_name`](https://github.com/fast-jsonapi/fast_jsonapi/blob/1d7c18f5dac0177a0a4d8728d84b4fd63b2d91df/lib/fast_jsonapi/object_serializer.rb#L278) is not referencing it at all either:

```ruby
def compute_serializer_name(serializer_key)
  return serializer_key unless serializer_key.is_a? Symbol
  namespace = self.name.gsub(/()?\w+Serializer$/, '')
  serializer_name = serializer_key.to_s.classify + 'Serializer'
  (namespace + serializer_name).to_sym
end
```

It looks like the polymorphic hash "works" when writing the relationships `type`/`id` hash.  It will compare the record with the classes specified in the hash, and then output the value of the key as the `type`.  However, when including records, it selects the serializer based on the record's class name... which is how the relationships work when `polymorphic` is simply set to `true`.  Therefore, if someone had a reason to specify `polymorphic` as a hash, and they are including records, then the types will likely mismatch.

Personally, I think that `polymorphic: true` is helpful in the typically simple cases: it infers the type and serializer from the name of the relationship.  However, I think we can design a better API for determining the correct type and serializer for polymorphic relationships and other occasions where they are not constant.  I think #29 is a better approach to solving this problem.

### PR DESCRIPTION
**Pull request by christophersansone**
**Overview** 

This PR implements the following changes:

1. The ability to specify a relationship's `serializer` option as a `Proc`, when the serializer may be dynamic or requires some custom logic.  Resolves #29.  Example:

```ruby
  class ZooSerializer
    include FastJsonapi::ObjectSerializer

    has_many :animals, serializer: -> (object) do
      if object.is_a?(Mammal)
        MammalSerializer
      elsif object.is_a?(Bird)
        BirdSerializer
      else
        AnimalSerializer
      end
    end
  end
```

2. When `record_type` is not specified on a relationship, the outputted `type` is now always determined from its serializer's `set_type`.

3. When `record_type` is not specified on a relationship, the relationship's serializer must always exist (so that it can get the `type`).

**Is it a breaking change for anyone?**

There are two edge cases where this can result in different behavior than before:

1. A `record_type` is not specified, and the old way to infer the type does not match that of the relationship's serializer.  In this case, this was likely invalid JSON:API output (see #30), so this change now prevents this from occurring.

2. A `record_type` is not specified, and the relationship's serializer is not defined.  This should not really be happening in practice... if a relationship cannot be serialized, there is no sense in defining the relationship to begin with.

**TL;DR**

The new serializer option behaves just like the `Proc` definitions for attributes and conditionals.  This gives the developer full control over determining the serializer to use based on any possible use case.

Because the `Proc` may output different serializers, and different serializers means having different `type` values, I needed a way to get the correct types somehow.  I chose to use the serializer's `set_type` value to offer a Single Source of Truth.  This resolves some uncertainty as to what `record_type` does, whether it chooses the serializer, what `type` is being outputted, etc (#14).  It also resolves #30, where the `type` in the `relationships` hash may differ from the `type` in the `included` list.  With this change, the `type` will always be consistent and simplifies the API because the developer no longer needs to define it in both the serializer and on all relationships with via `record_type`.

There were very few changes that I had to make to the existing specs... but there were changes.  I want to specifically point them out and encourage you to review the spec changes.  I annotated the changeset with comments for further clarity as well.

1.  Because the `type` is now resolving differently, it is possible that the output may change for some people.  When the object itself is serialized, as `data` or within `included`, the `type` always comes from the serializer's `set_type`.  That has not changed at all.  What has changed is that the `relationships` hash now uses that same value from `set_type`, as opposed to inferring it by the property name.  Therefore, the only cases where the output would be different after this change than before this change is if the relationships hash contained a different `type` than the record itself, which to me is a bug in the output.

2. As a result of the previous point, it now requires a serializer to be defined for each class being outputted, so that it can properly resolve the `type` key.  In other words, previously, you could define a relationship without actually having a serializer defined for those items, and it would infer the `type` from the property name.  I can’t imagine that this is an actual problem in practice, because it does no good to have a relationship defined without somehow being able to serialize those records.  I only mention it because there were some specs that defined relationships without defining the serializers.

**How do these changes impact performance?**

I tried to compare the rspec performance benchmarks between the `master` branch and this branch.  When the difference is microseconds, it is really hard to get perfectly consistent readings on my computer (late 2019 MacBook Pro).  Initial load appears to be consistently affected because of the up front serializer resolution, but we are talking about 0.3ms.  After that, performance is quite similar.  The tests with larger recordsets were often faster with the new implementation, though it is not consistent enough to state one way or the other.  I am confident saying that _the performance impact is negligible on initial render and at least as fast on subsequent renders._

**What should happen to `record_type`?**

The `record_type` option is still usable and will override the default behavior of looking up the serializer.  To me, the negatives outweigh the positives at this point.  It adds to the API surface and overlaps other functionality, causing confusion.  And worse, because it can result in the `type` in the `relationships` hash being different than the serialized record itself, it can result in invalid output.  I believe it should be deprecated, but it still "works" for now.

### CODE REVIEW

File: spec/shared/examples/object_serializer_class_methods_examples.rb [hunk 1]
```diff
@@ -2,17 +2,17 @@
   it 'returns correct relationship hash' do
     expect(relationship).to be_instance_of(FastJsonapi::Relationship)
     # expect(relationship.keys).to all(be_instance_of(Symbol))
-    expect(relationship.serializer).to be serializer
+    ######expect(relationship.serializer).to be serializer
     expect(relationship.id_method_name).to be id_method_name
-    expect(relationship.record_type).to be record_type
+    #####expect(relationship.record_type).to be record_type
```
Comment (christophersansone) [inline spec/shared/examples/object_serializer_class_methods_examples.rb]: I commented these out temporarily but forgot to fix them.  This is testing internal behavior that changed, so I need to figure out how to modify these tests.  Will hopefully do that tonight.

File: spec/lib/object_serializer_class_methods_spec.rb [hunk 2]
```diff
@@ -16,35 +16,51 @@
     end
 
     context 'with namespace' do
+
```
Comment (christophersansone) [inline spec/lib/object_serializer_class_methods_spec.rb]: The only material change in this file is that the serializer classes are actually defined.

File: spec/lib/object_serializer_inheritance_spec.rb [hunk 3]
```diff
@@ -51,6 +51,15 @@ class UserSerializer
     has_one :photo
   end
 
+  class Address
```
Comment (christophersansone) [inline spec/lib/object_serializer_inheritance_spec.rb]: These classes were referenced but not actually defined.  The serializer classes are now required.

File: spec/lib/object_serializer_inheritance_spec.rb [hunk 4]
```diff
@@ -93,7 +107,7 @@ class EmployeeSerializer < UserSerializer
     attributes :location
     attributes :compensation
 
-    has_one :account
+    has_one :account, serializer: EmployeeAccountSerializer
```
Comment (christophersansone) [inline spec/lib/object_serializer_inheritance_spec.rb]: This here was extremely unclear what the behavior should have been.  There is a property called `account`, but the class is defined as `EmployeeAccount`.  Should it assume `AccountSerializer` because of the property name?  Should it read the object and see that it's an `EmployeeAccount` and therefore use `EmployeeAccountSerializer`?

FYI, the existing behavior (and the updated behavior) is that `has_one :account` will resolve to `AccountSerializer`.  This has not changed.  What has changed is that it now requires the serializer to exist.  But, the reality is that you probably won't have an `AccountSerializer` when the class is `EmployeeAccount`.  To me, this looks like a bug here in the `EmployeeSerializer`, where we should expect it to be `EmployeeAccountSerializer` and therefore the relationship would somehow need to specify the serializer.

**Comment by christophersansone** [comment]
@stas Hey, just wondering if you've had a chance to consider this one.  I believe there are some important improvements here moving forward.  I'd love to help address any concerns you may have.

File: spec/lib/object_serializer_inheritance_spec.rb (hunk 4 above)
Comment (stas) [reply spec/lib/object_serializer_inheritance_spec.rb]: It looks a bit messy, but it seems like the `account` serializer was never defined and actually not used except in two tests (lines 144-151). I think we can just replace the account with something that already has a defined serializer, like move the `has_one :photo` into the `EmployeeSerializer` and update the tests to reflect `photo` instead of the `account`. Plus we delete a couple of extra lines.

File: lib/fast_jsonapi/object_serializer.rb [hunk 5]
```diff
@@ -314,7 +309,13 @@ def validate_includes!(includes)
             relationships_to_serialize = klass.relationships_to_serialize || {}
             relationship_to_include = relationships_to_serialize[parsed_include]
             raise ArgumentError, "#{parsed_include} is not specified as a relationship on #{klass.name}" unless relationship_to_include
-            klass = relationship_to_include.serializer.to_s.constantize unless relationship_to_include.polymorphic.is_a?(Hash)
+            if relationship_to_include.static_serializer
+              klass = relationship_to_include.static_serializer
+            else
+              # the serializer may change based on the object (e.g. polymorphic relationships),
+              # so inner relationships cannot be validated
+              break
```
Comment (stas) [inline lib/fast_jsonapi/object_serializer.rb]: Let's add a test for this use-case. We might need to actually just skip (`next`) instead of cancelling the validation of the remaining records (or I'm missing the context here, please correct/help me understand).

File: lib/fast_jsonapi/relationship.rb [hunk 6]
```diff
@@ -120,5 +153,74 @@ def run_key_transform(input)
...
+        serializer_for_name(serializer_name)
+      end
+    end
+
+    def serializer_for_name(name)
+      @serializers_for_name[name] ||= compute_serializer_for_name(name)
+    end
+
+    def compute_serializer_for_name(name)
+      namespace = owner.name.gsub(/()?\w+Serializer$/, '')
```
[the remark this answers is not in the document]
Comment (stas) [reply lib/fast_jsonapi/relationship.rb]: Would be great to actually use some namespaced fixtures for exactly this purpose.

[the remark this answers is not in the document]
Comment (christophersansone) [reply lib/fast_jsonapi/relationship.rb]: Long story short, I simply moved [compute_serializer_name](https://github.com/fast-jsonapi/fast_jsonapi/blob/c305b699a970ba60e77ba5d7d1f962390bcdc363/lib/fast_jsonapi/object_serializer.rb#L278) from the serializer into the relationship.  And because the relationship is now responsible for computing the serializer name when necessary, and the computation is based on the original serializer's name (for namespacing), we need a reference to the `owner` for this computation.

With this PR, the Relationship's `serializer` property is specifically the `serializer` parameter passed into the `belongs_to` / `has_many` definition.  Before this change, `serializer` was always specified on the Relationship, because if it was not defined, it was computed by the serializer and passed in as `serializer`, and the Relationship could not tell the difference.  This caused confusion for `polymorphic` relationships, because it didn't know whether to use the `serializer` or compute the `polymorphic`.

This way, the Relationship is now solely responsible for determining the serializer for each object in the relationship being serialized (Single Responsibility Principle).

Now, it is all clear.  If `serializer` is specified as a parameter, it will always be used, regardless of `polymorphic`.  If `polymorphic` is specified without `serializer`, it will compute the serializer name.  If neither are defined, it will resolve the _static serializer_ and only compute it once.

As for _how_ the name is computed, I did not change that at all.  I tend to think this PR has enough change as it is.  I think that if we want to address that, we should do so as a separate PR and address it holistically.

File: lib/fast_jsonapi/relationship.rb [hunk 7]
```diff
@@ -120,5 +153,74 @@ def run_key_transform(input)
...
+      serializer_name = name.to_s.demodulize.classify + 'Serializer'
```
[the remark this answers is not in the document]
Comment (christophersansone) [reply lib/fast_jsonapi/relationship.rb]: Again, I just moved this method.  [Here](https://github.com/fast-jsonapi/fast_jsonapi/blob/c305b699a970ba60e77ba5d7d1f962390bcdc363/lib/fast_jsonapi/object_serializer.rb#L281) is the same line before, doing what it is currently doing.  If we want to change the implementation, let's do it as a separate PR.

File: lib/fast_jsonapi/object_serializer.rb (hunk 5 above)
Comment (christophersansone) [reply lib/fast_jsonapi/object_serializer.rb]: This was also reported and fixed in #13.  Personally I would prefer to remove this method all together.  All it does is pre-validate the `include` option prior to serialization.  To me, this is not helpful at all because the serialization will fail anyway.  It is extra upfront work (and therefore a performance hit) every time a serialization occurs, and it cannot even be validated if the serializer is dynamic.  I think we are better off just assuming the `include` option is valid and let the serializer fail normally if not.

Before this change, it would incorrectly throw an exception for dynamic serializers (e.g. `polymorphic`).  I minimized the change while avoiding breakage by simply `break`ing instead of failing.

If we were to write a test case for this scenario, the correct test case would simply prove that this validation does not work correctly with dynamic serializers.  I would rather solve the problem a different way.

Maybe we can move the discussion over to #13 or start a different issue.  In the meantime, I would like to just have this bug fixed.

File: lib/fast_jsonapi/object_serializer.rb [hunk 8]
```diff
@@ -245,18 +245,20 @@ def create_relationship(base_key, relationship_type, options, block)
...
           name: name,
           id_method_name: compute_id_method_name(
             options[:id_method_name],
             "#{base_serialization_key}#{id_postfix}".to_sym,
             polymorphic,
+            options[:serializer],
             block
           ),
-          record_type: options[:record_type] || run_key_transform(base_key_sym),
+          record_type: options[:record_type],
```
[the remark this answers is not in the document]
Comment (christophersansone) [reply lib/fast_jsonapi/object_serializer.rb]: Yes, I created #34 to discuss deprecating `record_type`.  For now, I wanted to keep this PR's impact to a minimum.  We can't remove it without deprecating it because it will be a breaking change.  For now, if `record_type` is specified, it should be used, period.  That is how it worked before, so that is how it should remain until we remove the option all together.

**Comment by christophersansone** [comment]
I agree that the tests and fixtures can use some improvement.  I opened #33 to discuss having more, clearer tests for each relationship option, but it sounds like you have some ideas to broaden it even further.  I will defer to you on how to better structure everything related to tests, but I am open to whatever makes sense.

File: lib/fast_jsonapi/relationship.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (stas) [reply lib/fast_jsonapi/relationship.rb]: @christophersansone do you mind if we move the call to the `compute_serializer_for_name` into the `create_relationship`, and pass directly the computed serializer to `options[:serializer] || compute_serializer_for_name(name, owner)`?!

Also ideally, could we make this `compute_serializer_for_name` a public API method pretty please?
The idea is to allow pragmatically an user implement a _renderer_ instead of requiring to provide the serializer class name every time. What do you think?

A real-world rails renderer example could be found here:
https://github.com/stas/jsonapi.rb/blob/3ebf4982243db36d7b5a54ab3c733f76f1df71ba/lib/jsonapi/rails.rb#L100

File: lib/fast_jsonapi/relationship.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (stas) [reply lib/fast_jsonapi/relationship.rb]: Alright, we'll have to deal with it later then. Added #42

File: lib/fast_jsonapi/relationship.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (stas) [reply lib/fast_jsonapi/relationship.rb]: Also here: https://github.com/Netflix/fast_jsonapi/issues/449

File: lib/fast_jsonapi/object_serializer.rb (hunk 5 above)
Comment (stas) [reply lib/fast_jsonapi/object_serializer.rb]: Alright, let's do it in the next PR.

File: lib/fast_jsonapi/relationship.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (christophersansone) [reply lib/fast_jsonapi/relationship.rb]: @stas I can move it, but the relationship must know the difference between when `options[:serializer]` was specified vs. when the serializer name was computed, which it can't do if we are passing in `options[:serializer] || compute...`.  There is enough complexity to the order of operations that need to occur, that in my opinion, it should be the sole responsibility of `Relationship` to determine the serializer for a given relationship / object.  `Relationship` also needs `compute_serializer_for_name` at the object level for `polymorphic: true` relationships.

If your desire here is to simply make `compute_serializer_for_name` a public, static, overridable method on `ObjectSerializer`, I can do that no problem.  Then the relationship can simply call `owner.class.compute_serializer_for_name()` (or whatever) as needed.  Does that address your concern?

File: lib/fast_jsonapi/object_serializer.rb (hunk 8 above)
[the remark this answers is not in the document]
Comment (stas) [reply lib/fast_jsonapi/object_serializer.rb]: Alrighty, fair, let's do it separately.

File: lib/fast_jsonapi/relationship.rb [hunk 9]
```diff
@@ -120,5 +153,74 @@ def run_key_transform(input)
         input.to_sym
       end
     end
+
+    def initialize_static_serializer
```
Comment (stas) [inline lib/fast_jsonapi/relationship.rb]: @christophersansone I'm trying to understand the reasoning behind all this caching and extra attributes. Correct me if I'm wrong, but a relationship will have just one serializer klass, which we can calculate upon the relationship initialization DSL (say as part of the `create_relationship`). Same is true for the polymorphic case. So why do we need the extra logic here, seems overly complicated?

File: lib/fast_jsonapi/relationship.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (stas) [reply lib/fast_jsonapi/relationship.rb]: > I can move it, but the relationship must know the difference between when `options[:serializer]` was specified vs. when the serializer name was computed

I'm not sure I understand, why is this important?

We have these cases here:
 * [ ] a class was provided as an option, no computation is needed
 * [ ] a symbol/string was provided, we need to compute a klass out of it
 * [ ] a proc was provided, no computation is needed, will be called directly
 * [ ] nothing was provided, we use the relationship name and compute it as it would be case nr.2

All seem to be independent from the relationship state, the relationship only needs to know about the class or the proc it needs to use/call, same for the polymorphic option, I just think this will greatly simplify things.

> If so, is `compute_serializer_for_name` the right name for this public method?

Sounds good, we can even make it shorter and just use `serializer_for`. Thank you.

File: lib/fast_jsonapi/relationship.rb (hunk 9 above)
Comment (christophersansone) [reply lib/fast_jsonapi/relationship.rb]: > Correct me if I'm wrong, but a relationship will have just one serializer klass...

This is true in some cases but not all.  A relationship's serializer can either be _static_ or _dynamic_.

When it is static, it is determined once up front and will not change.  This is the most common use case:

```ruby
belongs_to :actor
has_many :movies
belongs_to :genre, serializer: CustomGenreSerializer
```

The serializer is considered _dynamic_ when it may change based on `object` or `params`.  This occurs in the following ways:

```
has_many :movies, polymorphic: true
has_many :actors, polymorphic: { ComedyActor => :custom_comedy, DramaActor => :custom_drama }
has_many :genres, serializer: -> (object, params) {
  ...
}
```

Therefore, the role of `initialize_static_serializer` is to determine _if_ the relationship's serializer can be static, and if so, which serializer it is.  It determines this once and then caches it forever.  If it determines that the serializer is dynamic, then it returns `nil`, stating that there is no static serializer.

Then, if there is no static serializer, it knows that it needs to dynamically determine the serializer.  In some cases, like `polymorphic: true`, it simply looks at the object class, computes the serializer by name once and memoizes it in `@serializers_for_name`.

The bottom line is this.  If `options[:serializer]` is specified, that takes precedence over everything.  The developer is specifically stating which serializer to use (or how to determine it if a Proc is specified).  If not, then it needs to infer it, based on whether `options[:polymorphic]` is specified, by the relationship name, etc.

Is it complicated?  Yes.  Is it overly complicated?  I don't think so, given that it all needed to be backwards compatible with `options[:record_type]` and `options[:polymorphic]`.  I basically just consolidated all the logic into one place.  Now to me the flow is quite simple:

1. Up front, it determines if the serializer is static or dynamic.
2. For every record being serialized, the relationship calls `serializer_for` to determine which serializer to use.  If the serializer is static, it returns the static serializer immediately.  If not, it performs the dynamic check (which also does what it can to memoize the results for performance purposes).

To get an idea of how much more consolidated the logic is, look in the original implementation about how many checks for `polymorphic` are all over the place.  Look at how many times it needs to iteratively call `serializer.to_s.constantize` or `object.class.name.demodulize.underscore.to_sym.to_s.constantize`.  And consider how much simpler the `Relationship` API is for determining the serializer of any given object.

Finally, to address the second part of your question:

> ...which we can calculate upon the relationship initialization DSL (say as part of the create_relationship)...

If the serializer is indeed static, then it is indeed determined "up front".  It is not done in `create_relationship` for two reasons:

1. As stated earlier, if `options[:serializer]` is specified, then it takes precedence over any other serializer determination such as `polymorphic`.  Therefore, we don't want to create the relationship with its serializer defined as `options[:serializer] || compute_serializer()`.  We _could_ still compute it in `create_relationship`, but it would have to be as a separate variable.  To me this is more complicated, not less, and at that point I would rather have the relationship handle it.

2. Before, the `serializer` passed into `Relationship` was a symbol, which still has to be constantized.  And it was constantized often.  Now, the static serializer determined by the relationship is the actual serializer class, and THAT is memoized.  Because of this, we cannot necessarily determine this when the Relationship is created, because the serializer class may not be loaded yet.  Therefore, it is determined "up front", i.e. the first time it is attempted.

Hopefully this answers all your questions.

File: lib/fast_jsonapi/relationship.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (christophersansone) [reply lib/fast_jsonapi/relationship.rb]: Other cases not mentioned above:

* When `polymorphic` is `true` and `serializer` is not specified
* When `polymorphic` is `true` and `serializer` is specified
* When `polymorphic` is a Hash and `serializer` is not specified
* When `polymorphic` is a Hash and `serializer` is specified

I know you feel this is overly complicated and you keep mentioning how a different implementation will "greatly simplify things", but I'm not sure I follow how.  Can you offer some pseudocode to illustrate your point?

[the remark this answers is not in the document]
Comment (stas) [reply lib/fast_jsonapi/relationship.rb]: @christophersansone Here's what I have in mind:
```ruby
Relationship.new(
  ...
  serializer: self.serializer_for(options[:serializer]) unless polymorphic.present?,
  polymorphic: polymorphic.transform_values { |_pv| self.serializer_for(_pv) },
  ...
)

def serializer_for(alias)
  # Will return either a class or a proc or raise some name error
end
```

Assuming we don't need to compute the serializer class if the polymorphic mapping is specified though...

This way we just need  to either use the already computed class or call the proc as part of the relationship.

[the remark this answers is not in the document]
Comment (christophersansone) [reply lib/fast_jsonapi/relationship.rb]: A couple concerns I see with the pseudocode:

1. `Relationship.new` is being called on load, which means that the resulting class of `serializer_for` must either be already loaded or can autoload.  Will that work in all cases?  Usually this type of work is deferred until first use, e.g. ActiveRecord's `belongs_to :user, class_name: 'UserSerializer'`.

2. What happens if both `polymorphic` and `serializer` is specified?  In your logic, `polymorphic` will take precedence.  Granted, I tend to think that both shouldn't be used together, but if they are, I feel like if `options[:serializer]` is specified, it should take precedence in all cases.

3. Ultimately, what does the code look like when a relationship is being serialized / included?  In what other areas of the code do we need to check `if polymorphic.present?` or if `serializer` is a Proc, and if so, execute it?  How often would this be replicated throughout the codebase?

I personally don't think this is any simpler.  I think if you followed your pseudocode to its logical conclusion, you would end up with something remarkably like my implementation. :smile:  My implementation simply encapsulates all of the logic of determining the serializer to use for any given record and exposes a nice clean `Relationship.serializer_for(object, params)`, and it resolves and caches as much as it can up front to ensure that subsequent calls are as fast as possible.

Trying to be helpful here but I can't keep splitting hairs on minor differences of opinion on how a line of code should be written. :smile:  I can move `Relationship.compute_serializer_for_name` to `ObjectSerializer.serializer_for`.  After that, can we just merge it?  From there, you are free to try to simplify it, and you would be able to do a more thorough comparison of the changes you want to make, make sure the tests pass, compare performance, etc.  Does that sound okay?

[the remark this answers is not in the document]
Comment (stas) [reply lib/fast_jsonapi/relationship.rb]: @christophersansone sorry was really just trying to understand the implementation not trying to argue about how it was done. Let's do it as you suggest, and get this merged. We can refine alter on.

**Comment by mmkarim** [comment]
Hi @christophersansone, 
Thank you for coming up with this feature. I was trying the fixes from your branch and found an unusual bug,

I declare a` has_many` relation with `polymorphic: true` for `documents`

```
has_many :documents, polymorphic: true, serializer: Serializer::Document
```

And when I check the response, the `id` of the `document` was showing the `id` of the `user`.
Here's the response,
 
```
    "data": {
        "id": "1043",
        "type": "user",
        "relationships": {
            "documents": {
                "data": {
                    "id": "1043",
                    "type": "document"
                }
            }
        }
    }
```
Hopefully it will help you to fix the bug.

**Comment by christophersansone** [comment]
@mmkarim Thanks for the report!  Can you confirm that this is not occurring in the master branch here?

I noticed some things in the existing code along the lines of properly resolving the IDs.  It seems like `id_method_name` is not being used correctly in all cases.  Maybe related to [this commit](https://github.com/fast-jsonapi/fast_jsonapi/commit/1d7c18f5dac0177a0a4d8728d84b4fd63b2d91df).

If a Pet `belongs_to :owner`, then `id_method_name` should refer to the way to retrieve the owner's ID from the pet model, e.g. `:owner_id`.  I think there is an issue where it is using `id_method_name` on the child instead of the parent.  I think `id_method_name` can be misinterpreted two different ways: "the foreign key name of the relationship" vs. "the primary key name of the related record".  I think it's being used in both ways, which is not correct.  I did not go down that rabbit hole in this PR.

FWIW, `id_method_name` should refer to the foreign key name.  If the primary key name needs to be determined from the relationship, it should look up the serializer and use `set_id`.  This could not be determined effectively prior to this PR, but now we are always looking up the serializer for relationships.  So, I think we can solve the root issue effectively once this PR is merged.

If you can confirm that this problem is in `master`, would you mind creating a separate issue?

**Comment by mmkarim** [comment]
@christophersansone I tried with the master branch. It's working correctly there.

**Comment by christophersansone** [comment]
@mmkarim Sorry I'm not quite sure.  I will need your help to investigate further.

I think what's happening is this.  You are specifying a "static" serializer for all records: `Serializer::Document`.  You are also specifying `polymorphic: true`.  The `polymorphic` option causes `id_method_name` to be `:id` rather than `document_ids`.

I think this is what's going on:

1. You are specifying a specific serializer: `Serializer::Document`.  This will be used for every child record, therefore it is considered a "static" serializer.

2. You are specifying `polymorphic`, which affects `id_method_name`.  When `polymorphic` is not specified, then by default it assumes it can fetch the IDs via `user.document_ids`.  When `polymorphic` is specified, it assumes that it must reach into each individual record and touch `id`, i.e. `user.documents.map(&:id)`.  As a result, `id_method_name` becomes `id` instead of `document_ids`.

3. During serialization, the static serializer is causing it to fetch the IDs from the foreign key, i.e. `user.document_ids`, but because `polymorphic` changed the name, it is looking up `user.id` instead.

I'll try to make this backward compatible with the old behavior, in case others run into this.  The right solution here is this:

* Let the Relationship determine what `id_method_name` should be if it is not specified in `options`.
* When the serializer is static, it should assume it can do the foreign key lookup, i.e. `user.document_ids`.
* When the serializer is dynamic (polymorphic or a Proc), it should iterate over the objects and use the ID field specified by its serializer's `set_id`.

Lastly, can you help me understand what you are expecting to accomplish when both `polymorphic` and `serializer` are specified?  Before this PR, `polymorphic` would take precedence over `serializer`, which means that it would attempt to look up different serializers for each object.  After this PR, `serializer` takes precedence over `polymorphic`, so it will always use `Serializer::Document`.  Which one are you actually trying to accomplish?  Is there a reason why you needed to specify both in this case?

As a rule of thumb, specifying `polymorphic` and `serializer` together is not a good idea because they have overlapping functionality, causing expectations to be unclear. @stas and I have discussed the impact of these two quite a bit.

**Comment by christophersansone** [comment]
@mmkarim cc @stas I just created #45 to address the overall issues with `id_method_name`.

**Comment by christophersansone** [comment]
@stas I created `ObjectSerializer#serializer_for` as we discussed.  Let me know what else is needed to merge!

@mmkarim I believe I fixed your issue by giving `polymorphic` precedence over `serializer`.  I would have preferred not to, but it does a better job of maintaining backward compatibility when both are specified.  Can you please confirm that this is fixed?

**Comment by mmkarim** [comment]
Amazing work @christophersansone. It's working accurately now. 👍

**Comment by jopotts** [comment]
Hi @christophersansone - great work. @mmkarim it would be good to have a test to show the failures, and prevent future regression. Not sure how easy that would be though.

And about @stas's comment about the fixtures and improved tests, I agree but as @christophersansone said, it should be separate from this PR. https://github.com/fast-jsonapi/fast_jsonapi/issues/33 is the start of that.

**Comment by ivan-garanin** [comment]
Also it will be great opportunity to specify `id_method_name` / `foreign_key`  in `Proc` too. As example 
```
  id_method_name: -> (object) do
      if object.is_a?(Mammal)
        :slug
      elsif object.is_a?(Bird)
        :uuid
      else
        :id
      end
    end
``` 
I have got some custom id's (`set_id`) in several serializers and it's a headache for me right now.

**Comment by stas** [comment]
@christophersansone I'd like to get this merged, please let me know if you're ok me picking up the work where you left. Ty :bowing_man:

**Comment by christophersansone** [comment]
@stas So sorry for dropping the ball.  I had some time over the holidays but have been swamped since.  I've been using it in production since then and it's been working great.  I did find a minor race condition that I resolved, but other than that, it hasn't changed.

I resolved the merge conflict here in Github (just the README).  I thought it would just create a new commit in my branch, but apparently not.  I would have just rebased instead had I known.  Anyway, I think it's good to go.

Is there anything else that needs to be done prior to merging?  Feel free to ask or to just take the ball and run with it, whatever works best!

**Comment by stas** [comment]
Thanks @christophersansone :bowing_man:

**Comment by stas** [comment]
@everlast240, actually I'd be happy to review an example and maybe put it into a test case if you're willing to share it.

@christophersansone thanks for helping us understand the internals :bow:
